vignettes/parallel.Rmd

```diff
@@ -95,8 +106,110 @@ This should return a vector of `TRUE` values if everything is appropriately setu
+
+## Benchmarks
+
+A simulated data set was tuned over a small grid of 5 candidate models and model performance was measured using 10-fold cross-validation. This means that a total of 50 models were created during tuning. 
+
+The computer is an iMacPro with 10 Intel chips running R version 4.2.0 (2022-04-22). Note that, with Intel chips, [hyper-threading](https://en.wikipedia.org/wiki/Hyper-threading) enables twice the number of simultaneous processes to be run at once. In the plots below, maximum CPU usage corresponds to 20 CPUs. The [Syrupy](https://github.com/jeetsukumaran/Syrupy) python library was used to monitor CPU usage. 
+
+### Computational methods
+
+The computations were run in a few different ways: 
+
+* __Completely sequential processing__. The idea was to have the h2o server use a single thread to process the models. The file `sequential.R` used `h2o.init(nthreads = 1)` to do this. Implicitly, `h2o.grid()` has a default that models are processed sequentially. 
+
+* __Multithreaded processing__: The server was configured to use all CPUs on the host via `h2o.init(nthreads = -1)` while `h2o.grid()` is still set to sequential processing. This code is in `multithreaded.R`. 
+
+* __Multithreaded parallel processing__: Along with `h2o.init(nthreads = -1)`, the calls to the h2o server used `h2o.grid(parallelism = 5)` so that a maximum of 5 models (i.e. the entire grid) could be processed at once. 
+
+* __Multithreaded, multicore parallel processing__: Along with `h2o.init(nthreads = -1)`, the calls to the h2o server used `h2o.grid(parallelism = 50)`. In additional, multicore parallel processing via the foreach and doMC packages were used to send all of the candidate models for all resamples to the server at once. 
+
+* __Multithreaded, PSOCK parallel processing__: Along with `h2o.init(nthreads = -1)`, the calls to the h2o server used `h2o.grid(parallelism = 50)`. Here, parallel processing was enabled via the foreach and doParallel packages (using a PSOCK cluster) to achieve similar results
+
+
+### Results
+
+When the baseline configuration of single threaded, sequential processing was used the execution time for the grid search was 296.6 seconds. The pattern of CPU usage was:
+
+<img src="figure/completely-sequential-1.svg" alt="plot of chunk completely-sequential" style="display: block; margin: auto;" />
+
+The 10 clusters of high utilization correspond to the h2o server processing the 5 candidate models for each of the 10 resamples. CPU utilization is about 1, as expected. 
+
+Once multiple threads were allowed, the grid search lasted 311 seconds (slightly slower than the baseline). Looking at the `Rout` file, the output lists that
+
+```
+    H2O cluster total cores:    20 
+    H2O cluster allowed cores:  20 
+```
+
+so it is unclear why the processing was relatively slow. The usage graph: 
```

Comment:
It's definitely not what I'd expect from H2O. 

Possible reason could be the usage of hyper-threading. I'm not sure if all but at least some Intel processors have one floating point unit per core and with HT you can have 2 parallel threads on one core which means that in floating point intensive workload the threads will have to wait for each other.

Another possible reason related to HT is security fixes - couple years ago there were several security issues related to HT (Meltdown, Spectre,...) and one of the mitigation techniques was to disable HT altogether so I assume it might take some performance hit on some workloads.

For both cases it could help to use `nthreads=10`.

Another reason could be cache utilization - the more threads, the more cache invalidations => more time spent on  waiting on memory. 

It can also be related to how we split the data, if the dataset is small, it could very well make the training slower with higher parallelism (more time spent on communication and synchronization).  Trying with bigger data could make the more parallel version perform better in this case.

Anyway, now I'm curious about it so I'll try to run the [benchmark](https://github.com/topepo/agua-h2o-benchmark) with a profiler. If I find some reason I will mention it here as well.

Reply:
I did not find the definition of `sim_classification` function but if it takes as its first argument number of rows to generate, I would say the reason for slower parallel run is really just small data (10k rows).

With 10-fold cv it will use 9k rows to train the model and if we have 20 threads we will process 450 rows per thread so the time spent on communication/synchronization might be significant when compared to the computation time and cause this behavior.

@ledell do we have any recommendation about how many threads should we use based on the dataset size? I know we have some heuristic for [GLM](https://docs.h2o.ai/h2o/latest-stable/h2o-docs/data-science/glm.html#faq) (`nodes = rows*columns^2/(nthreads*1e8)`) but I don't know if we have something like that in general.

My general recommendation would be to use the H2O parallelism (using `nthreads` and/or h2o cluster) for bigger data (but I don't know where the threshold is).

Reply:
@tomasfryda Recommendation is to use default `nthreads` except if strong reason not to do so (e.g. running on laptop and don't want h2o to use all my cpus). The heuristic that you mentioned is to optimize the number of `nodes` as they need to be defined beforehand, but once a node is started, it's up to H2O to optimize its behaviour according to the number of threads/cpus available, not the other way round.